### PR TITLE
Generate `THIRD-PARTY-NOTICES.txt` for the VMR

### PR DESCRIPTION
**Pull request by premun**
Adds a new command `darc vmr generate-tpn` that generates the `THIRD-PARTY-NOTICES.txt` file which is a compilation of other `THIRD-PARTY-LICENSE` files from the product repos.

### CODE REVIEW

File: src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrCommandLineOptions.cs [hunk 1]
```diff
@@ -3,25 +3,47 @@
...
 
 internal abstract class VmrCommandLineOptions : CommandLineOptions
 {
-    [Value(0, Required = true, HelpText = 
-        "Repository names in the form of NAME or NAME:REVISION where REVISION is a commit SHA or other git reference (branch, tag). " +
-        "Omitting REVISION will synchronize the repo to current HEAD. Pass 'all' to update all repositories.")]
-    public IEnumerable<string> Repositories { get; set; }
-
-    [Option("vmr", Required = false, HelpText = "Path to the VMR; defaults to git root above cwd.")]
+    [Option("vmr", Required = true, HelpText = "Path to the VMR; defaults to git root above cwd.")]
```
Comment (andriipatsula) [inline src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrCommandLineOptions.cs:20]: `; defaults to git root above cwd` this sounds unclear for me. Could we rephrase this a little bit.

File: src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/ThirdPartyNoticesGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        ILogger<ThirdPartyNoticesGenerator> logger)
+    {
+        _vmrInfo = vmrInfo;
+        _dependencyTracker = dependencyTracker;
+        _fileSystem = fileSystem;
+        _logger = logger;
+    }
+
+    /// <summary>
+    /// Generates the THIRD-PARTY-NOTICES.txt file by assembling other similar files from the whole VMR.
```
Comment (andriipatsula) [inline src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/ThirdPartyNoticesGenerator.cs:42]:
```suggestion
    ///     Generates the THIRD-PARTY-NOTICES.txt file by assembling other similar files from the whole VMR.
```

Comment (andriipatsula) [reply src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/ThirdPartyNoticesGenerator.cs:42]: I noticed you updated comments in several places... to be compliant with previous changes

File: src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/ThirdPartyNoticesGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            {
+                paths.Add(tpn);
+            }
+        }
+
+        return paths.OrderBy(p => p);
+    }
+
+    /// <summary>
+    /// Reads the header of the THIRD-PARTY-NOTICES file (until 2 empty lines are found).
```
Comment (andriipatsula) [inline src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/ThirdPartyNoticesGenerator.cs:99]:
```suggestion
    ///     Reads the header of the THIRD-PARTY-NOTICES file (until 2 empty lines are found).
```

File: src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/ThirdPartyNoticesGenerator.cs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+            }
+
+            header.Add(line);
+        }
+
+        return header;
+    }
+
+    /// <summary>
+    /// Checkes if a given path is a THIRD-PARTY-NOTICES.txt file.
```
Comment (andriipatsula) [inline src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/ThirdPartyNoticesGenerator.cs:123]:
```suggestion
    ///     Checkes if a given path is a THIRD-PARTY-NOTICES.txt file.
```

**Review by andriipatsula — APPROVED** [review]

File: src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/ThirdPartyNoticesGenerator.cs (hunk 2 above)
Comment (premun) [reply src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/ThirdPartyNoticesGenerator.cs:42]: Yeah, there's this one file that has these indented but honestly I'd prefer to not indent. I only kept that one file in the same shape.

File: src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrCommandLineOptions.cs (hunk 1 above)
Comment (premun) [reply src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrCommandLineOptions.cs]: Fixed

**Review by andriipatsula — APPROVED** [review]

**Pull request merged**